Implement pagination in github package

puerco opened the pull request:
#### What type of PR is this?
/kind bug
/kind feature 

#### What this PR does / why we need it:

The first commit in this PR adds proper pagination support to functions in the `github` package that were missing it.

It works by adding a new Options struct to the `github` package to configure the size of pages we get from calls to GitHub.

The next PR alters the following functions to paginate according to the page size defined in `Options.ItemsPerPage`:

* ListBranches()
* ListReleaseAssets()
* TagExists()
* LatestGitHubTagsPerBranch()


#### Which issue(s) this PR fixes:

Ref: https://app.slack.com/client/T09NY5SBT/CJH2GBF7Y/thread/CJH2GBF7Y-1607502787.026300

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?

```release-note
- The GitHub package now has a basic options struct to control how it behaves
- Add `ItemsPerPage` option to the `github` pkg to control the size of pages we ask from github
- Add pagination when calling the GitHub API from the following functions: `ListBranches() ListReleaseAssets() TagExists() LatestGitHubTagsPerBranch()`

```

hasheddan commented on pkg/github/github.go:
```diff
@@ -792,18 +810,25 @@ func (g *GitHub) ListReleaseAssets(
-	if err != nil {
-		return exists, errors.Wrap(err, "listing repository tags")
-	}
+	options := &github.ListOptions{PerPage: 100}
```
Is `100` the max value allowed here? Should we allow consumers of this package to limit the number of results returned if they don't need all pages?

puerco replied:
100 is the maximum per page, yes.
Of the three functions I'm patching, two return lists: `ListAssets` and `ListBranches`. The number of assets is always small, I don't think you would need to limit the returns.

In the case of branches, perhaps you would need to limit the returns if you had hundreds and hundreds of branches. Would it make sense?

cpanato replied:
I think we can have a default value if the `PerPage` is not set by the client as we did in another pkg (https://github.com/cpanato/release/blob/master/pkg/vulndash/adapter/adapter.go#L88-L90)

justaugustus replied:
+1 for a default value and configuration options.

saschagrunert reviewed (approved):
/lgtm
/hold
for @hasheddan's discussion point

puerco commented:
@hasheddan I've added a new `Options` struct to the github object as we have in the new krel code to make it configurable. I've reworked the pagination in calls that get lists from GitHub to use the new configurable option. These functions are now aware of the `ItemsPerPage` option:

* ListBranches()
* ListReleaseAssets()
* TagExists()
* LatestGitHubTagsPerBranch()

hasheddan commented on pkg/github/github.go:
```diff
@@ -432,7 +473,7 @@ type TagsPerBranch map[string]string
 	// List tags for all pages
 	allTags := []*github.RepositoryTag{}
-	opts := &github.ListOptions{PerPage: 100}
+	opts := &github.ListOptions{PerPage: g.Options().ItemsPerPage}
```
Just noting here that we could have a panic if someone had manually constructed the `*Github` struct. All of our `New___` functions set the default options, so in practice this likely will not be a huge issue, but worth pointing out. One solution for areas like this are to add a `GetItemsPerPage()` method to `*Options` as Go allows you to call methods on nil structs :)

puerco replied:
I've added the accessor method to the options struct and changed all calls to use it :+1:

hasheddan reviewed (approved):
Apologies for the long delay here @puerco! I aded one note but LGTM, please feel free to lift hold when you are ready 👍 

/lgtm

hasheddan commented:
@puerco ping :)

/retest

puerco commented:
/test pull-release-image-kube-cross

hasheddan commented:
@puerco looks like we need rebase here

puerco commented:
I was working in the GitHub pkg and remembered I had to fix this PR. PTAL @hasheddan :)

cpanato reviewed (approved):
/lgtm

saschagrunert reviewed (approved):
/lgtm
/hold cancel
